### PR TITLE
BE-46: HashQL: Register all variables mentioned during inference

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

Traverse a type to register any encountered generic variables. This is required so that in the case we have variables that are just not hit, we still record a diagnostic telling us to do so.

As a drive-by, it fixes the README diagram to be more readable.

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

### CODE REVIEW

File: libs/@local/hashql/core/src/type/environment/infer.rs [hunk 1]
```diff
@@ -156,6 +166,8 @@ impl<'env, 'heap> InferenceEnvironment<'env, 'heap> {
     }
 
     pub fn collect_constraints(&mut self, variance: Variance, subtype: TypeId, supertype: TypeId) {
+        // TODO: we need to add a way that "tops" out, meaning that is called from outside, and one
+        // from inside, outside also registers all variables.
```
[the remark this answers is not in the document]
Comment (indietyp) [reply libs/@local/hashql/core/src/type/environment/infer.rs]: was a remnant from a previous integration attempt

File: libs/@local/hashql/hir/tests/ui/lower/inference/collect-variable.jsonc [hunk 2]
```diff
@@ -0,0 +1,10 @@
+//@ run: fail
+//@ description: Collect variables that aren't explicitely mentioned
```
[the remark this answers is not in the document]
Comment (indietyp) [reply libs/@local/hashql/hir/tests/ui/lower/inference/collect-variable.jsonc]: fixed in [c6b897e](https://github.com/hashintel/hash/pull/7793/commits/c6b897e1c23e44a6ea5d80dc28c41055f2d96187)

File: libs/@local/hashql/core/src/type/environment/infer.rs [hunk 3]
```diff
@@ -149,6 +149,16 @@ impl<'env, 'heap> InferenceEnvironment<'env, 'heap> {
         variable
     }
 
+    pub fn add_variables(&mut self, variables: impl IntoIterator<Item = Variable>) {
```
Comment (hashdotai) [inline libs/@local/hashql/core/src/type/environment/infer.rs:152]: This method should include a clear comment explaining its purpose. From the context, it seems this is used to register variables that might not be used in a unification constraint, but it's not clear from the implementation alone.

File: libs/@local/hashql/hir/src/lower/inference.rs [hunk 4]
```diff
@@ -123,6 +129,9 @@ impl<'env, 'heap> TypeInference<'env, 'heap> {
         TypeInferenceResidual<'heap>,
         Vec<TypeCheckDiagnostic>,
     ) {
+        let variables = self.collector.take_variables();
```
Comment (hashdotai) [inline libs/@local/hashql/hir/src/lower/inference.rs:132]: It looks like you're accessing the collector as a mutable reference, but then using `take_variables()` which consumes the collector. This means you can no longer use the collector after this point. It might be better to either:

1. Clone the variables instead of taking them
2. Add a method to get a reference to the variables
3. Document this behavior clearly if it's intentional

File: libs/@local/hashql/hir/tests/ui/lower/inference/collect-variable.jsonc [hunk 5]
```diff
@@ -0,0 +1,10 @@
+//@ run: fail
+//@ description: Collect variables that aren't explicitly mentioned
```
Comment (hashdotai) [inline libs/@local/hashql/hir/tests/ui/lower/inference/collect-variable.jsonc:2]: Consider adding a test for the edge case where we have a type with variables that don't participate in any constraints. This test is a good start, but additional test cases with more complex scenarios would help ensure the robustness of this feature.

File: libs/@local/hashql/hir/tests/ui/lower/inference/collect-variable.stderr [hunk 6]
```diff
@@ -0,0 +1,14 @@
+[type-check::unconstrained-type-variable] Error: Unconstrained type variable
+   ╭─[ <unknown>:6:19 ]
+   │
+ 6 │   "Dict<Integer | _, String>",
+   │                   ┬  
+   │                   ╰── Cannot infer type for this variable - no usage constraints available
```
Comment (hashdotai) [inline libs/@local/hashql/hir/tests/ui/lower/inference/collect-variable.stderr:6]: You're updating the error message here, but it would be helpful to include a more specific diagnostic for cases where variables are mentioned but not constrained. Something like "Variable mentioned in type but not constrained by usage" would be more informative than the generic "Cannot infer type for this variable".

File: libs/@local/hashql/hir/src/lower/inference.rs (hunk 4 above)
Comment (indietyp) [reply libs/@local/hashql/hir/src/lower/inference.rs:132]: no.

File: libs/@local/hashql/core/src/type/environment/infer.rs (hunk 3 above)
Comment (indietyp) [reply libs/@local/hashql/core/src/type/environment/infer.rs:152]: There's a comment in the implementation, also tracked in https://linear.app/hash/issue/BE-64/hashql-create-docs-for-publicly-accessible-environment-methods

File: libs/@local/hashql/hir/tests/ui/lower/inference/collect-variable.stderr (hunk 6 above)
Comment (indietyp) [reply libs/@local/hashql/hir/tests/ui/lower/inference/collect-variable.stderr:6]: ?
